### PR TITLE
Merge #17: Add descaling page

### PR DESCRIPTION
**Pull request by LightArrowsEXE**
Wrote a post centered around how and when to descale, as well as covering how to go about it.

### CODE REVIEW

File: encoding/descaling.md
Comment (cN3rd) [inline encoding/descaling.md]: Misses an emphasis on "high relative errors using/given by the script".
It might confuse beginners who have no prior knowledge of descaling.

File: encoding/descaling.md
Comment (cN3rd) [inline encoding/descaling.md:161]: This is also the place to discuss what are considered to be "low relative errors" and "high relative errors", when to ignore them, and how to interpret them and the graph.

File: encoding/descaling.md
Comment (cN3rd) [inline encoding/descaling.md]: Emphasis on that you're looking for a low error **spike**.
As you can see in the graph, the closer you get to the full resolution of the source (e.g. 1080p) the smaller the "relative error" is.

File: encoding/descaling.md
Comment (cN3rd) [inline encoding/descaling.md]: An appendix or a small section about popular kernels would be useful for beginners.

File: encoding/descaling.md
Comment (cN3rd) [inline encoding/descaling.md:329]: Reversing the lines shows that getnative outputs height and you work around that:

```Py
height = 878
width = vsutil.get_w(height)
```

Comment (LightArrowsEXE) [reply encoding/descaling.md:329]: Changed

File: encoding/descaling.md
Comment (LightArrowsEXE) [reply encoding/descaling.md:161]: Expanded greatly on how to read the graphs, as well as showing some examples.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: I don’t like how this sounds. “You may have heard before that most…” or similar.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “There are many circumstances under which…”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: comma after example

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: isn’t blurriness just… blur?

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Please use “might have”. This isn’t commie.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:1]: General thing: search and replace `'` with proper apostrophes (`’`).

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “will yield better results than a simple resample”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “dabble” means “to work without any deep involvement” (MW), but specific masks imply deep involvement. It’s also “dabble with”.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “as long as”. So long as is [dying out](https://books.google.com/ngrams/graph?content=as+long+as%2Cso+long+as&year_start=1800&year_end=2008&corpus=6&smoothing=5&share=&direct_url=t1%3B%2Cas%20long%20as%3B%2Cc0%3B.t1%3B%2Cso%20long%20as%3B%2Cc0#t1%3B%2Cas%20long%20as%3B%2Cc0%3B.t1%3B%2Cso%20long%20as%3B%2Cc0) and may even feel unnatural to some readers.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “need for” feels more natural to me, even though [both seem to be correct](https://english.stackexchange.com/questions/6623/need-of-vs-need-for).

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:58]: Can we not use a hyperlink here directly?

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: It’s not just designed *around* that, it’s designed *for* this one purpose. It does nothing else.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: There *are* going to be differences (plural)

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:68]: 1. Proper quotes `“”`.
2. Even though the source isn’t perfect, you can still find perfect settings (the exact same settings the studio used), so this sounds misleading to me.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:87]: I’d prefer “resizing” to make it less repetitive with “descaling”.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Ideally,

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Lower what chance? The sentence isn’t really coherent.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Here are

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: no comma here

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: maybe “check“ instead of “look in”?

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “where you executed the script”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: comma after checked

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Any low value is a low relative error. Higher resolutions tend to have lower errors because less information is lost or changed during de- and rescaling. We need to make clear that the spikes break this pattern by having a much lower error rate compared to similar resolutions, not compared to all tested resolutions.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “sidenote … to note” sounds repetitive

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:181]: Add an example of a descale graph for a native 1080p frame.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “pit*fall* you can *fall* in”? :thinking:

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: There are

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Don’t use your wrapper just to import something in these examples. A regular lsmash call will do just fine.
Also remove the `r` before the string as it’s not needed when using `/` as a path separator.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: It seems odd to me that we’re talking about de*scaling*, re*scaling*, down*scaling* but now up*sampling*. I’d just call this link upscale, even if it links to upsampling to keep terminology consistent within this section.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: There are multiple splines. Kekkai Sensen used Spline16 iirc, so list 16 and 36 separately.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: There might be a word missing here. Or there’s one too many.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Why is there a linebreak after each word?

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: For bicubic,

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:301]: None of your examples follow the second rule, and I don’t know of any that do. I think it’s either `b + 2c = 1` or one of them is 0 and the other could be anything. We don’t need 4 rules to express that.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: repeated “this”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:312]: “descaled requires floating point input” what that is should be explained elsewhere, and we don’t want to update this page if another float format is added in the future.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: You only use `core` from vapoursynth, so you can just `from vapoursynth import core`

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Variable names should be in lowercase

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Please put the `#comment` above the corresponding function call.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “the more likely it is”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “that you can use”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:358]: Magnification using nearest neighbor. Using bilinear or bicubic on the preview won’t show you anything of value.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: You’re just discarding all of these clips without ever saying why. The reader can’t know that you’re changing the `kgf.join()` call between comparisons.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md:394]: You’re not supposed to call methods prefixed with `_`. Use `mask_descale` instead.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “or trying”

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Don’t make it sounds so good or people will think it magically fixes everything. Like descaling to the wrong resolution and then having the detail mask and eedi carry you. :smug_nep:

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: Make it more clear that rescaling/rpow2 magic is optional. Currently, this implies that you have to upscale again.

File: encoding/descaling.md
Comment (kageru) [inline encoding/descaling.md]: “Most, if not all, relevant”

**Review by kageru — CHANGES_REQUESTED** [review]
Note: All of my QC comments are to be understood as suggestions, not absolute, objective truth.

File: encoding/descaling.md
Comment (LightArrowsEXE) [reply encoding/descaling.md:394]: For this example I wanted to generate the mask specifically. As far as I can tell, I can't do that with `mask_descale`.

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md]: What's the line where `might've` isn't accepted but `you've` is? Is there something semi-official or just your gut feeling? (Imo this is fine. Has been adressed already, though.)

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md:1]: Do we want to make this a guideline? I don't particularly care about the correct use of apostrophes in a context like this.

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md:58]: Hyperlinks in text are disrupting, artificially increase line length in source and work against semantic linefeeds. It's better to declare them after the paragraph (or at the end if they are used multiple times).

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md:68]: Good point on 2. For 1. see my comment earlier on apostrophes.

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md:87]: ^nit but valid

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md:312]: i.e. reference to the docs in a footnote?

**Comment by FichteFoll** [comment]
Reviewed @kageru's review and left comments where I disagreed or am indifferent. Comments I upvoted I support strongly, the others just normally. I also resolved those that have been addressed already.

I'll trust kageru's judgement for this section and wait until his review has been addressed, after which I'll do a final pass and most likely merge.

---

@cN3rd, you can combine multiple comments in a single review if you click on "Start a review" on the first comment, proceed to add more comments and finally click on the "Review Changes" button to submit your review comments together.

File: encoding/descaling.md
Comment (kageru) [reply encoding/descaling.md]: Just gut feeling. You’ve feels… acceptable(?) to me, but might’ve gives me this feeling of forced contraction just for the sake of it.

File: encoding/descaling.md
Comment (kageru) [reply encoding/descaling.md:394]: Nvm, I’m an idiot. I thought I put a debug parameter in there.

File: encoding/descaling.md
Comment (kageru) [reply encoding/descaling.md:1]: Maybe we should get a third opinion on this. I like them, but I guess most people don’t care.

File: encoding/descaling.md
Comment (kageru) [reply encoding/descaling.md:58]: Will this syntax insert a hyperlink on the word `getnative` using the link defined below?

File: encoding/descaling.md
Comment (LightArrowsEXE) [reply encoding/descaling.md]: Leaving a note for myself to look at this later, as I can't think of a clear way to do this yet. To me, the "spike" itself seems like an obvious enough example of the pattern being broken. Will think about it.

File: encoding/descaling.md
Comment (LightArrowsEXE) [reply encoding/descaling.md]: I was under the impression that you always had to break the line after a comma with semantic linefeeds.

**Comment by LightArrowsEXE** [comment]
These two comments remain unresolved, as I don't know how to exactly go about them:

https://github.com/Irrational-Encoding-Wizardry/guide.encode.moe/pull/17/files/2718ea784f5a17ab79dfd7e8766b5a103723c2a5#diff-90c73b01b7f851b1c04c0f924d50b478
https://github.com/Irrational-Encoding-Wizardry/guide.encode.moe/pull/17/files/2718ea784f5a17ab79dfd7e8766b5a103723c2a5#diff-90c73b01b7f851b1c04c0f924d50b478

(also I just realized I said "discord" instead of "github" in the commit description)

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md:58]: Yes, this is a very standard markdown feature and used throughout the guide already.

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md]: The oxford comma only applies for enumerations. Is a comma actually valid here?

File: encoding/descaling.md
Comment (FichteFoll) [reply encoding/descaling.md]: You don't *have* to do this if the enumeration is short, like it is here. Dave just has a dendency to do it for all of them, but I'm a believer of applying common sense in these situations.

File: encoding/descaling.md
Comment (FichteFoll) [inline encoding/descaling.md]: You need to update all these upper-case references (also applies to the following lines).

**Review by FichteFoll — CHANGES_REQUESTED** [review]
Went through the review again and resolved the addressed feedback. Here's one more, however.

File: encoding/descaling.md
Comment (kageru) [reply encoding/descaling.md]: This is not an Oxford comma. You always put a comma between two independent clauses joined by a conjunction. “I did A, and I did B” needs a comma. “I did A and B” does not have a comma because “and B” is not an independent sentence.

**Review by FichteFoll — APPROVED** [review]

File: encoding/descaling.md
Comment (kageru) [reply encoding/descaling.md:358]: nearest neighbor in lowercase

**Comment by OrangeChannel** [comment]
I've sent a [PR](https://github.com/LightArrowsEXE/guide.encode.moe/pull/2) to @LightArrowsEXE , addressing some of kageru's suggestions, and also making sure visually the captions are properly formatted and actually display (due to blank lines not surrounding the attachment line). There are still a few issues to address, namely

- https://github.com/Irrational-Encoding-Wizardry/guide.encode.moe/pull/17#discussion_r298801336
- https://github.com/Irrational-Encoding-Wizardry/guide.encode.moe/pull/17#discussion_r298801844

Re: the fancy quotes discussion. It is much easier to type `'` , and `"` and is a universal standard according to [the unicode character table](https://unicode-table.com/en/sets/quotation-marks/), but not sure if that's an official source or anything. It seems that the basic quotes are used in all of the other pages, and would require a decent amount of replacing and then probably a mention in the contributing section (these aren't easy to type on a default US keyboard) if we were to change this.

edit: I've also checked that the page and all the links/attachments work properly on a live version of the book, so it should be all good to go once these two issues are addressed.

**Comment by FichteFoll** [comment]
I agree that using fancy quotes is probably too much for a guide like this. I already voiced my specitcism, I believe. Added a review to the mentioned PR (https://github.com/LightArrowsEXE/guide.encode.moe/pull/2).

I believe we can merge once those are addressed.

**Comment by OrangeChannel** [comment]
Bringing this discussion back over here, https://github.com/LightArrowsEXE/guide.encode.moe/pull/2#discussion_r298813839, so I can get others' opinions on this.

`but imo it makes sense to have this be a direct hyperlink, assuming the kagefunc library has docs. I wonder if the glossary would do us any good here because it is going to be in backticks most of the time.`

This is definitely a possibility, would likely require taking kagefunc out of the in-line code as hopefully that doesn't count as a glossary use (will test before commiting). Regarding the glossary entry, what should be put under it besides a hyperlink i.e. https://encode.moe/kagefunc ? @kageru any suggestions?

**Comment by kageru** [comment]
I don’t think we need to write a full glossary entry for Python modules or plugins. A reader can follow the hyperlink(s) and read the readme/docs/docstrings of the module, so writing our own description feels like redundant work to me.

**Comment by FichteFoll** [comment]
The idea with the glossary is that it will insert the links (to the glossary entry) automatically as soon as the word is spotted in use. The alternative would be to manually add a link to every occurance of a module name (but with the advantage of putting it in backticks, i.e. `` [`kagefunc`][kagefunc]``.

**Comment by FichteFoll** [comment]
I'll merge this for now. We can decide on script linking later.

**Pull request merged**
